### ISSUE TITLE
【2026犀牛鸟开源人才专属】【中高难度】MoE 优化专项：MoE 专家剪枝与动态超参数调度优化

### ISSUE THREAD

**Issue by isLinXu** [Tencent/YOLO-Master#52]
在 COCO 或 VisDrone 数据集上训练 YOLO-Master-EsMoE-N 模型
使用 MoEPruner 对训练好的模型进行专家剪枝，对比至少 5 组阈值（threshold ∈ {0.05, 0.10, 0.15, 0.20, 0.30}）
对比两种恢复策略：剪枝后直接推理 vs 剪枝后 + LoRA 微调 10-epoch 恢复
测量并记录每个实验点的：mAP50-95、mAP50、FLOPs（通过 get_gflops()）、Latency（ms）、Params（M）、每层保留专家数、专家利用率 Gini 系数
绘制 "阈值 → mAP / FLOPs / Latency" 三维曲线和 Pareto 前沿图（精度 vs 延迟），标注推荐的 "Sweet Spot"
设计并实现至少一种动态超参数调度策略（如根据 expert_usage 的 Gini 系数自动调整 balance_loss_coeff，或根据验证 mAP 饱和情况动态退火 top_k），需给出调度公式并论证合理性
与固定参数基线进行对照实验（至少 3 组：基线 + 动态调度 + 消融组），计算收敛加速比（实验组达到基线最终精度 95% 所需 epoch 比例）
分析动态调度的副作用（如训练不稳定、最终 mAP 下降），提出改进建议
提出场景化推荐：如服务器端推荐阈值 0.10、边缘端推荐阈值 0.20，并附数据支撑
完成后，在 GitHub Discussion 发表技术总结文章并提供实验脚本仓库链接
动态调度策略代码修改需保持向后兼容，可提交 Pull Request
https://github.com/Tencent/YOLO-Master
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/nn/modules/moe/pruning.py
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/nn/modules/moe/diagnostics.py
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/nn/modules/moe/analysis.py
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/nn/modules/moe/loss.py
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/nn/modules/moe/modules.py
https://github.com/Tencent/YOLO-Master/blob/main/scripts/compare_moe_coco128.py
https://github.com/Tencent/YOLO-Master/blob/main/scripts/bench_moe_micro.py
难度：中

**Comment by vankari** [comment]
已认领本任务

**Comment by ShadowBearKill** [comment]
已认领本任务

**Comment by 2021113691** [comment]
已认领本任务

**Comment by nicholasdezai** [comment]
已认领本任务

**Comment by Dayuxiaoshui** [comment]
已认领本任务

**Comment by Ricky-7-Yan** [comment]
已认领本任务

**Comment by sensg233** [comment]
已认领本任务

**Comment by Lfan-ke** [comment]
已认领本任务

**Comment by shizuku-desir7e** [comment]
已认领本任务

**Comment by SidKC** [comment]
已认领本任务

**Comment by SidKC** [comment]
我在原帖和已有工作的基础上完成了一轮独立复现，覆盖专家剪枝、LoRA10 恢复、Gini 动态调度及三组对照实验。复现过程中还发现并修复了两个会影响结果可信度的问题：路由利用率只保留最近一次快照，以及 recovery 后同一个 epoch 被重复计入调度状态。

## 实验设置

- 数据集：VisDrone，训练集/验证集分别为 `6471/548` 张图；
- 剪枝阈值：`0.05 / 0.10 / 0.15 / 0.20 / 0.30`；
- 恢复策略：剪枝后直接推理、剪枝后 LoRA 微调 10 epochs；
- 动态对照：Fixed Baseline、Gini Dynamic、Fixed-Low Ablation；
- 延迟测试：同一单卡硬件类别、FP32、batch 1、640×640，仅测模型 forward；
- 动态三组使用相同 seed、数据、batch、输入尺寸和初始模型状态。

正式校准使用训练集的确定性 10% 子集，seed 为 42，实际选中并观测 `647/647` 张图、81 batches、32527 instances。剪枝信号与输入 checkpoint 的 SHA-256 绑定，同时检查实际观测样本、非空标签和剪枝后的物理结构。

## 专家剪枝结果

五档阈值最终只形成两种物理结构：

- `0.05–0.20`：均未发生实际剪枝，专家数保持 `3/3/3/3`；
- `0.30`：专家数变为 `2/2/3/3`。

| 方案 | 专家结构 | mAP50-95 | mAP50 | P95 forward | Params | GFLOPs |
| --- | --- | ---: | ---: | ---: | ---: | ---: |
| Dense baseline | `3/3/3/3` | 0.203575 | 0.350382 | 13.379 ms | 2.653 M | 8.49 |
| Threshold 0.30 direct | `2/2/3/3` | 0.123546 | 0.242145 | 13.210 ms | 2.625 M | 7.89 |

阈值 `0.30` 的 mAP50-95 相对下降约 39.3%，P95 model-forward 只缩短约 `0.169 ms`。按照预先设定的绝对 mAP50-95 降幅不超过 `0.01` 的质量门槛，本轮实验没有找到可接受的 Sweet Spot。

`.05–.20` 的逻辑实验结果仍保留在机器可读结果中，但它们对应相同的物理结构，因此没有作为四种不同的部署配置解释。

## LoRA10 恢复结果

仓库当前的原生 LoRA 路径会训练 adapters 和 detection head，本次结果按照这一实际训练范围记录。

| 方案 | Best mAP50-95 | Final mAP50-95 | P95 forward | GFLOPs |
| --- | ---: | ---: | ---: | ---: |
| Baseline LoRA10 | 0.181443 | 0.00491 | 22.103 ms | 10.81 |
| Pruned LoRA10 | 0.142967 | 0.00946 | 22.024 ms | 10.20 |

LoRA10 没有恢复到原始 dense baseline 的 `0.203575`，同时增加了推理计算量和延迟。两组在训练后期都出现明显质量坍塌，因此这里同时报告 best 和 final checkpoint。

这部分实验还发现了一个 MoLoRA checkpoint 状态问题：checkpoint 恢复了 `_step_count` buffer，但 Python 侧 `_step_count_cpu` 仍从 0 开始，导致 dynamic top-k warmup 被重新执行。对应修复已拆成独立 PR：

- [PR #110：恢复 checkpoint 中的 top-k warmup 状态](https://github.com/Tencent/YOLO-Master/pull/110)

## Gini 动态调度

动态组在每个 epoch 内累计全部 batch 的专家利用率，对逐层 Gini 取均值，然后更新 balance-loss 系数：

```text
ema_t = beta * ema_(t-1) + (1 - beta) * gini_t

coeff_(t+1) = clip(
    base_coeff * exp(alpha * (ema_t - target_gini)),
    min_balance_coeff,
    max_balance_coeff
)
```

路由更集中时提高 balance-loss 系数，路由较均衡时降低系数。该功能默认关闭。

调度采用 accepted-epoch 语义：只有 epoch 通过验证和 recovery 检查后才提交调度状态与 trace。被 recovery 拒绝的 epoch 会清空本轮累计，不推进 EMA、事件计数或 trace。

修正后的动态组严格生成 epoch `1..10` 共 10 条 trace，每轮包含：

```text
3236 = 4 × 809
```

次路由观测。最终事件计数为：

```text
opportunity_count = 10
event_count = 10
nontrivial_action_count = 10
```

对应修复已提交：

- [PR #109：修复 epoch 路由统计与 recovery 重复调度](https://github.com/Tencent/YOLO-Master/pull/109)

## 动态调度三组对照

| 实验组 | 平衡策略 | Best mAP50-95 | Best epoch | Final mAP50-95 |
| --- | --- | ---: | ---: | ---: |
| Fixed Baseline | 固定 `1.0` | 0.03551 | 4 | 0.00750 |
| Gini Dynamic | EMA 指数调度 | 0.04144 | 6 | 0.00905 |
| Fixed-Low Ablation | 固定 `0.3` | 0.03950 | 4 | 0.00493 |

动态组的平均/最终 Gini 为 `0.04615/0.01058`，平均/最终 balance-loss 系数为 `0.82399/0.80416`。

三组都出现了明显的 late-stage quality collapse。由于固定基线最终精度已经坍塌，三组都在 epoch 1 达到“基线最终精度 95%”，形式上的收敛比例均为 `1.0`，这个指标在当前实验中没有区分力。

Gini Dynamic 的 best 和 final mAP50-95 略高于固定基线，但目前只有单 seed、10-epoch 结果，暂时只能作为后续验证的候选现象，还不足以说明动态调度带来了稳定的收敛加速。

## 场景建议

- 服务器端：当前五档阈值中没有满足质量门槛的非平凡剪枝点，建议暂时保持 dense/no-op 配置；
- 边缘端：阈值 `.30` 的质量损失远大于当前 PyTorch forward 延迟收益，也不建议使用；
- `.10/.20` 当前没有产生物理剪枝，因此没有单独列为部署配置。

## 代码和复现材料

本轮补充了以下内容：

- 完整 epoch 路由利用率累计；
- recovery-safe 调度和 accepted-epoch trace；
- `opportunity/event/nontrivial-action` 事件计数；
- 剪枝信号与 checkpoint SHA-256 绑定；
- 剪枝后的专家、router 和 `top_k` 结构检查；
- 五档逻辑阈值到唯一物理结构的去重执行；
- 双顺序延迟 benchmark；
- 5×2 结果组装、三目标曲线和 Pareto 分析；
- 质量门槛内没有可行点时明确记录“未观察到 Sweet Spot”。

相关链接：

- [中文实验说明与结果汇总](https://github.com/SidKC/YOLO-Master/blob/rhino-2026/issue-52-moe-pruning-schedule/docs/moe_pruning_dynamic_schedule.md)
- [Issue #52 实验整合分支](https://github.com/SidKC/YOLO-Master/tree/rhino-2026/issue-52-moe-pruning-schedule)
- [PR #109：动态路由统计与 recovery 修复](https://github.com/Tencent/YOLO-Master/pull/109)
- [PR #110：MoLoRA checkpoint warmup 状态修复](https://github.com/Tencent/YOLO-Master/pull/110)
- [完整技术总结（Issue #111）](https://github.com/Tencent/YOLO-Master/issues/111)

实验整合分支的相关回归结果为：

```text
229 passed, 6 skipped
```

本轮实验没有得到适合直接部署的剪枝阈值，但厘清了五档逻辑阈值对应的实际物理结构，也保留了完整的负结果、测量脚本和可审查的 correctness 修复。下一步更适合先获得稳定的长周期训练基线，再进行多随机种子验证，并采用不依赖 collapsed-final checkpoint 的收敛目标。

欢迎维护者对当前 PR 拆分方式、稳定基线设置和后续实验优先级提出建议。

> 已认领本任务

**Comment by JiaCheng310** [comment]
已认领本任务

**Comment by hfu622** [comment]
已认领本任务

**Comment by SidKC** [comment]
进展更新：

作为 2026 腾讯犀牛鸟开源人才培养计划 Issue #52 的后续贡献，我重新基于最新 main 审计了此前实验中发现的 MoLoRA checkpoint warmup 状态问题。

当前生产实现已通过持久化 `_step_count` buffer 正确恢复 dynamic top-k warmup，不再需要重复提交旧 PR #110 的生产代码修复；但现有测试仍缺少完整的训练态 checkpoint round-trip 行为覆盖。

现已拆分提交一条聚焦的单文件回归测试：

- #118

该测试通过真实 training-mode forward 推进 warmup，并验证加载 `state_dict` 后 `_step_count` 和 effective top-k 均保持一致。测试可在 CPU 上运行，不修改生产代码。

**Comment by songhahaha66** [comment]
已认领本任务

**Comment by zhaozefeng67-blip** [comment]
已认领本任务

**Comment by Fe1ixFe1icis** [comment]
已认领本任务

**Comment by Mezca1** [comment]
已认领本任务

**Comment by Lfan-ke** [comment]
Discussion: https://github.com/Tencent/YOLO-Master/discussions/197

**Comment by meadows5241** [comment]
已认领本任务

**Comment by Abby-like-code** [comment]
已认领本任务

**Comment by WanNian-Li** [comment]
已认领本任务

**Comment by Pnofd** [comment]
已认领本任务

**Comment by isLinXu** [comment]
@SidKC @Lfan-ke @delei-kong @vankari 完成任务并获奖🏆🎉

**Comment by isLinXu** [comment]
@2021113691 @Ricky-7-Yan @Lfan-ke @vankari @delei-kong @SidKC已完成任务🌹
